workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/DialogRunner.kt

```diff
@@ -0,0 +1,151 @@
+package com.squareup.workflow1.ui.container
+
+import android.app.Dialog
+import android.content.Context
+import android.os.Bundle
+import android.os.Parcel
+import android.os.Parcelable
+import android.os.Parcelable.Creator
+import androidx.core.view.doOnAttach
+import androidx.core.view.doOnDetach
+import androidx.lifecycle.Lifecycle.Event.ON_DESTROY
+import androidx.lifecycle.LifecycleObserver
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.OnLifecycleEvent
+import com.squareup.workflow1.ui.Compatible
+import com.squareup.workflow1.ui.ViewEnvironment
+import com.squareup.workflow1.ui.WorkflowUiExperimentalApi
+import com.squareup.workflow1.ui.androidx.WorkflowLifecycleOwner
+import com.squareup.workflow1.ui.compatible
+
+@WorkflowUiExperimentalApi
+internal class DialogRunner<T : Overlay>(
+  initialRendering: T,
+  initialViewEnvironment: ViewEnvironment,
+  private val bounds: RectProvider,
```

Comment:
This seems to say:
the creator of a dialog runner will say where it should appear as opposed to:
- I will tell it the "cover area", the "anchor area", etc, and let the runner and its idiosyncrasy to decide where to appear depending on those inputs.

Reply:
As I see now, the particular implementations of overlays are not the dialog runners, but the overlay dialog factories. My comment still holds: These bounds might be cover-area, anchor-area and let the overlay dialog factory to update the final location depending on the inputs.

Something like:
```
Runner {
+ coverArea: RectProvider
+ anchorArea: RectProvider

  // when dialog is shown
  coverArea.onChange { coverRect = it }
  anchorArea.onChange { anchorRect = it }
  // when dialog is dismissed
  coverArea.onChange = null
  anchorArea.onChange = null
  
  protected coverRect: Rect by observable(...) { updateDialogLocation(...) }
  protected coverRect: Rect by observable(...) { updateDialogLocation(...) }
  fun updateDialogLocation() {
    someone has to decide how to calculate dialog position depending on cover + anchor
    that is not clear to me now, because Runner is fixed, and factory doesn't receiver cover + anchor
    I'd say that particular factories now the dialog they create
    and they also know what positioning they implement
    so they have the responsibility to update (or provide the cover+anchor -> bounds lambda at least).
  }
}
```

Reply:
Question is: what's the conceptual difference between a runner and a factory, unless the runner contains the common code (instead of having a base factory class, which might be a good idea).

Reply:
As I see now, the particular implementations of overlays are not the dialog runners, but the overlay dialog factories. My comment still holds: These bounds might be cover-area, anchor-area and let the overlay dialog factory to update the final location depending on the inputs.

Something like:
```kotlin
Runner {
+ coverArea: RectProvider
+ anchorArea: RectProvider

  // when dialog is shown
  coverArea.onChange { coverRect = it }
  anchorArea.onChange { anchorRect = it }
  // when dialog is dismissed
  coverArea.onChange = null
  anchorArea.onChange = null
  
  protected coverRect: Rect by observable(...) { updateDialogLocation(...) }
  protected coverRect: Rect by observable(...) { updateDialogLocation(...) }
  fun updateDialogLocation() {
    dialogOrNull
  }
}
```